Fix for #7809: check for use-site errors on attribute before synthesizing

jcouv opened the pull request:
Add checks on two paths in TrySynthesizeAttribute that could return an attribute even when use-site errors exist on that optional attribute's constructor or the named arguments for the attribute. 

The [corresponding VB code](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Symbols/WellKnownMembers.vb,139) does not have this issue.

CC @dotnet/roslyn-compiler @AlekseyTs for code review.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -300,7 +300,9 @@ internal static Symbol GetRuntimeMember(NamedTypeSymbol declaringType, ref Membe
         {
-            var ctorSymbol = (MethodSymbol)GetWellKnownTypeMember(constructor);
+            DiagnosticInfo useSiteDiagnostics = null;
+            var ctorSymbol = (MethodSymbol)Binder.GetWellKnownTypeMember(this, constructor, out useSiteDiagnostics, true);
```
Please use the last parameter as a named argument.

cston replied:
Are we reporting the `useSiteDiagnostics`?

AlekseyTs replied:
We are not supposed to report any diagnostics at this point.

jcouv replied:
@cston, the change only affects optional attributes. My understanding is that we just don't synthesize such optional attributes if there is an issue with them. 
That's also the behavior on the VB side.

I will double-check with @AlekseyTs to confirm desired behavior.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -323,7 +325,7 @@ internal static Symbol GetRuntimeMember(NamedTypeSymbol declaringType, ref Membe
                 foreach (var arg in namedArguments)
                 {
-                    var wellKnownMember = GetWellKnownTypeMember(arg.Key);
+                    var wellKnownMember = Binder.GetWellKnownTypeMember(this, arg.Key, out useSiteDiagnostics, true);
```
Please use the last parameter as a named argument.

cston commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -300,7 +300,9 @@ internal static Symbol GetRuntimeMember(NamedTypeSymbol declaringType, ref Membe
             ImmutableArray<KeyValuePair<WellKnownMember, TypedConstant>> namedArguments = default(ImmutableArray<KeyValuePair<WellKnownMember, TypedConstant>>))
         {
-            var ctorSymbol = (MethodSymbol)GetWellKnownTypeMember(constructor);
+            DiagnosticInfo useSiteDiagnostics = null;
```
Please rename to `diagnosticInfo` (singular). And the initializer is unnecessary.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Synthesized.cs:
```diff
@@ -1653,5 +1653,116 @@ class B : A
         }
+
+        [Fact, WorkItem(7809, "https://github.com/dotnet/roslyn/issues/7809")]
+        public void SynthesizeAttributeWithUseSiteErrorFails()
```
Let's add similar regression test for VB.

cston commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Synthesized.cs:
```diff
@@ -1653,5 +1653,116 @@ class B : A
+    {
+        public class DefaultMemberAttribute { }
+    }
+}
```
Are these needed: `ParamArrayAttribute`, `RuntimeFieldHandle`, `RuntimeTypeHandle`, `Activator`, `NotSupportedException`, `OutAttribute`, `DefaultMemberAttribute`?

jcouv replied:
@cston, I will verify and reduce to minimal set.

cston commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Synthesized.cs:
```diff
@@ -1653,5 +1653,116 @@ class B : A
+            // Build an mscorlib including String
+            var mslibComp = CreateCompilation(new string[] { mslib });
+            var diag = mslibComp.GetDiagnostics();
+            Assert.False(diag.Any(d => d.Severity == DiagnosticSeverity.Error));
```
Consider using `mslibComp.VerifyDiagnostics()` instead. Similar comment for other compilations below.

jcouv commented:
@cston, @AlekseyTs, I pushed a commit addressing code review feedback. It simplifies the code for fake mscorlib (a lot), adds the corresponding VB regression test, and addresses the other smaller points.

cston commented:
LGTM

jcouv commented:
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/4757/
// Retest reason: Error looks transient

AlekseyTs commented:
LGTM

jcouv commented:
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/4905/
// Retest reason: This is one of the flaky tests.
